### PR TITLE
Various fixes to SSL4EO sampler

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR contains the following improvements to our sampler script:

- [x] Don't ignore warnings (they're there for a reason)
- [x] Resume is boolean, not a file (save_path already tells us where the file is)
- [x] Populate R-tree at the same time as we read the resume file (no need to loop twice)
- [x] Delete the old sample file if it already exists (previously, indices_range of 0 10 followed by 0 100 would result in a single file with 110 entries, 10 duplicated, and with different, possibly overlapping bboxes for those 10)
- [x] Don't check ext_coords for every index (no longer allows sampling in reverse order, but we don't need that, and this is faster)
- [x] Save to sample file at the same time as R-tree population (allows resuming even if script is killed due to timeout)

I wanted to add `np.random.seed(0)` to the script, but that would result in the first `indices_range[1]` samples we draw already being present in the resume file, meaning we would just skip them. I don't know of an easy way to update the random seed to the prior state without saving it.

@wangyi111

### CODE REVIEW

**Review by wangyi111 — APPROVED** [review]

**Comment by wangyi111** [comment]
> I wanted to add np.random.seed(0) to the script, but that would result in the first indices_range[1] samples we draw already being present in the resume file, meaning we would just skip them.

Also related to https://github.com/microsoft/torchgeo/pull/1311#discussion_r1186218303. The problem here is that a fixed seed (0) is set for the random generator, so always get the same sequence of random numbers. We can use a different seed for different range arguments, i.e. pass the range arguments as input for the seed.

```
seed = hash((indices[0], indices[1]))
random.seed(seed)
```
or 
```
seed = hash((indices[0], indices[1])) % 2**32
np.random.seed(seed)
```
I guess this should work.

File: experiments/ssl4eo/sample_ssl4eo.py [hunk 1]
```diff
@@ -104,76 +98,61 @@ def create_bbox(
...
+            # Sample new coord and check overlap
+            while True:
+                # (lon,lat) of top-10000 cities
+                new_coord = sample_point(cities, args.std)
+                bbox = create_bbox(new_coord, bbox_size_degree)
+                if not list(rtree_coords.intersection(bbox)):
+                    break
+
+            rtree_coords.insert(i, bbox)
+            new_coords[i] = new_coord
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply experiments/ssl4eo/sample_ssl4eo.py:152]: I think you're right, I'll remove

**Comment by wangyi111** [comment]
> Will that work such that `--indices-range 0 10` followed by `--indices-range 10 20` produces the same values as `--indices-range 0 20`? That's what I want in terms of reproducibility.

That's probably not easy to do...

**Comment by adamjstewart** [comment]
That's what I figured. If someone really needs reproducibility, then can grab the bounds of each file with rasterio and use those to create the sampling list.

**Comment by calebrob6** [comment]
> Will that work such that --indices-range 0 10 followed by --indices-range 10 20 produces the same values as --indices-range 0 20? That's what I want in terms of reproducibility.

If you want this behavior you could always generate a list from 0 to `args.indices_range[1]` with a fixed seed and then grab the subset you need.

**Comment by adamjstewart** [comment]
Meh, not important.

**Review by wangyi111 — APPROVED** [review]

**Pull request merged**